[C++] Fails to parse std::function or segfaults.

### Information
ESMB version: `ESBMC version 7.5.0 64-bit x86_64 linux`
Kernel/OS: `Linux debian 6.1.0-17-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.69-1 (2023-12-30) x86_64 GNU/Linux`
GCC version: `gcc (Debian 12.2.0-14) 12.2.0`
Clang version: `Debian clang version 14.0.6 Target: x86_64-pc-linux-gnu`

### Code

```cpp
#include <cassert>
#include <functional>

int main()
{
    std::function<int(int, int)> sum = [](int a, int b) -> int
    {
        return a + b;
    };

    assert(sum(5, 7) == 12);

    return 0;
}
```

If we run `esbmc --cppstd 11 --input-file lambda_test.cpp` we get the following errors:
<details>

```shell
$ esbmc --cppstd 11 --input-file lambda_test.cpp 
ESBMC version 7.5.0 64-bit x86_64 linux
Target: 64-bit little-endian x86_64-unknown-linux with esbmclibc
Parsing lambda_test.cpp
In file included from lambda_test.cpp:2:
/tmp/esbmc-cpp-headers-b8f5-b0f8-f273/functional:15:54: error: no template named 'unary_function'
template<class Predicate> class unary_negate: public unary_function<
                                                     ^
/tmp/esbmc-cpp-headers-b8f5-b0f8-f273/functional:28:55: error: no template named 'binary_function'
template<class Predicate> class binary_negate: public binary_function<
                                                      ^
/tmp/esbmc-cpp-headers-b8f5-b0f8-f273/functional:43:51: error: no template named 'unary_function'
template<class Operation> class binder1st: public unary_function<
                                                  ^
/tmp/esbmc-cpp-headers-b8f5-b0f8-f273/functional:60:51: error: no template named 'unary_function'
template<class Operation> class binder2nd: public unary_function<
                                                  ^
/tmp/esbmc-cpp-headers-b8f5-b0f8-f273/functional:77:41: error: no template named 'unary_function'
class pointer_to_unary_function: public unary_function<Arg, Result> {
                                        ^
/tmp/esbmc-cpp-headers-b8f5-b0f8-f273/functional:90:42: error: no template named 'binary_function'
class pointer_to_binary_function: public binary_function<Arg1, Arg2, Result> {
                                         ^
/tmp/esbmc-cpp-headers-b8f5-b0f8-f273/functional:103:25: error: no template named 'unary_function'
class mem_fun_t: public unary_function<T*, S> {
                        ^
/tmp/esbmc-cpp-headers-b8f5-b0f8-f273/functional:115:26: error: no template named 'binary_function'
class mem_fun1_t: public binary_function<T*, A, S> {
                         ^
/tmp/esbmc-cpp-headers-b8f5-b0f8-f273/functional:127:31: error: no template named 'unary_function'
class const_mem_fun_t: public unary_function<T*, S> {
                              ^
/tmp/esbmc-cpp-headers-b8f5-b0f8-f273/functional:139:29: error: no template named 'unary_function'
class mem_fun_ref_t: public unary_function<T, S> {
                            ^
/tmp/esbmc-cpp-headers-b8f5-b0f8-f273/functional:151:30: error: no template named 'binary_function'
class mem_fun1_ref_t: public binary_function<T, A, S> {
                             ^
/tmp/esbmc-cpp-headers-b8f5-b0f8-f273/functional:163:35: error: no template named 'unary_function'
class const_mem_fun_ref_t: public unary_function<T, S> {
                                  ^
/tmp/esbmc-cpp-headers-b8f5-b0f8-f273/functional:320:37: error: no template named 'const_mem_fun1_t'; did you mean 'const_mem_fun_t'?
template<class S, class T, class A> const_mem_fun1_t<S, T, A> mem_fun(
                                    ^~~~~~~~~~~~~~~~
                                    const_mem_fun_t
/tmp/esbmc-cpp-headers-b8f5-b0f8-f273/functional:127:7: note: 'const_mem_fun_t' declared here
class const_mem_fun_t: public unary_function<T*, S> {
      ^
/tmp/esbmc-cpp-headers-b8f5-b0f8-f273/functional:320:37: error: too many template arguments for class template 'const_mem_fun_t'
template<class S, class T, class A> const_mem_fun1_t<S, T, A> mem_fun(
                                    ^                      ~~
/tmp/esbmc-cpp-headers-b8f5-b0f8-f273/functional:127:7: note: template is declared here
class const_mem_fun_t: public unary_function<T*, S> {
      ^
/tmp/esbmc-cpp-headers-b8f5-b0f8-f273/functional:322:9: warning: use of function template name with no prior declaration in function call with explicit template arguments is a C++20 extension [-Wc++20-extensions]
        return const_mem_fun1_t<S, T, A>(f);
               ^
/tmp/esbmc-cpp-headers-b8f5-b0f8-f273/functional:341:1: error: no template named 'const_mem_fun1_ref_t'; did you mean 'const_mem_fun_ref_t'?
const_mem_fun1_ref_t<S, T, A> mem_fun_ref(S (T::*f)(A) const) {
^~~~~~~~~~~~~~~~~~~~
const_mem_fun_ref_t
/tmp/esbmc-cpp-headers-b8f5-b0f8-f273/functional:163:7: note: 'const_mem_fun_ref_t' declared here
class const_mem_fun_ref_t: public unary_function<T, S> {
      ^
/tmp/esbmc-cpp-headers-b8f5-b0f8-f273/functional:341:1: error: too many template arguments for class template 'const_mem_fun_ref_t'
const_mem_fun1_ref_t<S, T, A> mem_fun_ref(S (T::*f)(A) const) {
^                          ~~
/tmp/esbmc-cpp-headers-b8f5-b0f8-f273/functional:163:7: note: template is declared here
class const_mem_fun_ref_t: public unary_function<T, S> {
      ^
/tmp/esbmc-cpp-headers-b8f5-b0f8-f273/functional:342:9: warning: use of function template name with no prior declaration in function call with explicit template arguments is a C++20 extension [-Wc++20-extensions]
        return const_mem_fun1_ref_t<S, T, A>(f);
               ^
lambda_test.cpp:6:10: error: no member named 'function' in namespace 'std'
    std::function<int(int, int)> sum = [](int a, int b) -> int
    ~~~~~^
lambda_test.cpp:6:26: error: expected '(' for function-style cast or type construction
    std::function<int(int, int)> sum = [](int a, int b) -> int
                      ~~~^
lambda_test.cpp:6:31: error: expected '(' for function-style cast or type construction
    std::function<int(int, int)> sum = [](int a, int b) -> int
                           ~~~^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
ERROR: PARSING ERROR
```

</details>

When running `esbmc --cppstd 11 --input-file lambda_test.cpp --no-abstracted-cpp-includes` we get a segfault:

<details>

```shell
$ esbmc --cppstd 11 --input-file lambda_test.cpp --no-abstracted-cpp-includes
ESBMC version 7.5.0 64-bit x86_64 linux
Target: 64-bit little-endian x86_64-unknown-linux with esbmclibc
Parsing lambda_test.cpp
Converting
Segmentation fault (Core dump written)
```

</details>

### Expected Result
ESBMC successfully compiles the file.

--- a/regression/esbmc-cpp11/lambda/Lambda--as-function-no-capture-no-inits-fail/main.cpp
+++ b/regression/esbmc-cpp11/lambda/Lambda--as-function-no-capture-no-inits-fail/main.cpp
@@ -1,0 +1,11 @@
+#include <cassert>
+#include <functional>
+
+int main()
+{
+  std::function<int(int, int)> sum = [](int a, int b) -> int { return a + b; };
+
+  assert(sum(5, 7) == 12222);
+
+  return 0;
+}

--- a/regression/esbmc-cpp11/lambda/Lambda--as-function-no-capture-no-inits-fail/test.desc
+++ b/regression/esbmc-cpp11/lambda/Lambda--as-function-no-capture-no-inits-fail/test.desc
@@ -1,0 +1,5 @@
+CORE
+main.cpp
+--std c++11
+
+^VERIFICATION FAILED

--- a/regression/esbmc-cpp11/lambda/Lambda--as-function-no-capture-no-inits/main.cpp
+++ b/regression/esbmc-cpp11/lambda/Lambda--as-function-no-capture-no-inits/main.cpp
@@ -1,0 +1,11 @@
+#include <cassert>
+#include <functional>
+
+int main()
+{
+  std::function<int(int, int)> sum = [](int a, int b) -> int { return a + b; };
+
+  assert(sum(5, 7) == 12);
+
+  return 0;
+}

--- a/regression/esbmc-cpp11/lambda/Lambda--as-function-no-capture-no-inits/test.desc
+++ b/regression/esbmc-cpp11/lambda/Lambda--as-function-no-capture-no-inits/test.desc
@@ -1,0 +1,5 @@
+CORE
+main.cpp
+--std c++11
+
+^VERIFICATION SUCCESSFUL$

--- a/regression/esbmc-cpp11/lambda/Lambda-no-capture-no-inits-fail/main.cpp
+++ b/regression/esbmc-cpp11/lambda/Lambda-no-capture-no-inits-fail/main.cpp
@@ -1,0 +1,6 @@
+#include <cassert>
+
+int main()
+{
+  assert([](int a) { return a + 2; }(2) == 444);
+}

--- a/regression/esbmc-cpp11/lambda/Lambda-no-capture-no-inits-fail/test.desc
+++ b/regression/esbmc-cpp11/lambda/Lambda-no-capture-no-inits-fail/test.desc
@@ -1,0 +1,5 @@
+CORE
+main.cpp
+--std c++11
+
+^VERIFICATION FAILED

--- a/regression/esbmc-cpp11/lambda/Lambda-no-capture-no-inits/main.cpp
+++ b/regression/esbmc-cpp11/lambda/Lambda-no-capture-no-inits/main.cpp
@@ -1,0 +1,6 @@
+#include <cassert>
+
+int main()
+{
+  assert([](int a) { return a + 2; }(2) == 4);
+}

--- a/regression/esbmc-cpp11/lambda/Lambda-no-capture-no-inits/test.desc
+++ b/regression/esbmc-cpp11/lambda/Lambda-no-capture-no-inits/test.desc
@@ -1,0 +1,5 @@
+CORE
+main.cpp
+--std c++11
+
+^VERIFICATION SUCCESSFUL$

--- a/src/clang-cpp-frontend/clang_cpp_convert.cpp
+++ b/src/clang-cpp-frontend/clang_cpp_convert.cpp
@@ -976,6 +976,34 @@ bool clang_cpp_convertert::get_expr(const clang::Stmt &stmt, exprt &new_expr)
     break;
   }
 
+  case clang::Stmt::LambdaExprClass:
+  {
+    const clang::LambdaExpr &lambda_expr =
+      static_cast<const clang::LambdaExpr &>(stmt);
+
+    if (lambda_expr.capture_size() > 0)
+    {
+      log_error("ESBMC does not support lambdas with captures for now");
+      return true;
+    }
+
+    get_struct_union_class(*lambda_expr.getLambdaClass());
+
+    // Construct a new object of the lambda class
+    typet lambda_class_type;
+    if (get_type(
+          *lambda_expr.getLambdaClass()->getTypeForDecl(), lambda_class_type))
+      return true;
+    new_expr = gen_zero(pointer_typet(lambda_class_type));
+    // Generating a null pointer only works as long as we don't have captures
+    // When we want to add support for captures, we have to probably call
+    // the lambda class constructor here
+    // See https://cppinsights.io/ which is a great tool to see how lambdas
+    // are translated to C++ code
+
+    break;
+  }
+
   default:
     if (clang_c_convertert::get_expr(stmt, new_expr))
       return true;

